File: 백준/Gold/3055. 탈출/탈출.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int r;
    static int c;
    static char[][] arr;
    static int[][] visited;
    static ArrayDeque<Point> q;
    static int endX;
    static int endY;
    static boolean canEnd = false; // 끝낼 수 있나?

    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {-1, 1, 0, 0};

    static class Point {
        int x;
        int y;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static void setInputs() throws IOException {
        st = new StringTokenizer(br.readLine());
        r = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());

        arr = new char[r][c];
        q = new ArrayDeque<>();
        for (int i = 0; i < r; i++) {
            String word = br.readLine();
            for (int j = 0; j < c; j++) {
                char c = word.charAt(j);
                arr[i][j] = c;

                if (c == '*') {
                    q.add(new Point(i, j));
                } else if (c == 'S') {
                    q.addFirst(new Point(i, j));
                } else if (c == 'D') {
                    endX = i;
                    endY = j;
                }
            }
        }
        visited = new int[r][c];
    }

    static void bfs() {
        while (!q.isEmpty()) {
            Point now = q.pop();
            int x = now.x;
            int y = now.y;

            if (arr[endX][endY] == 'S') {
                canEnd = true;
                return;
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx < 0 || ny < 0 || nx >= r || ny >= c) {
                    continue;
                }

                // 고슴도치가 다음 칸으로 이동할 때 . D 이면 이동 가능
                if (arr[x][y] == 'S' && (arr[nx][ny] == '.' || arr[nx][ny] == 'D')) {
                    arr[nx][ny] = 'S';
                    visited[nx][ny] = visited[x][y] + 1;
                    q.add(new Point(nx, ny));
                }

                // 물이 다음 칸으로 이동할 때 . S 이면 이동 가능 -> 물이 나중에 움직이니까 고슴도치 칸 잡아먹어도 됨.
                // 어차피 queue에는 고슴도치 칸이 들어있어서 괜찮음.
                else if (arr[x][y] == '*' && (arr[nx][ny] == '.' || arr[nx][ny] == 'S')) {
                    arr[nx][ny] = '*';
                    q.add(new Point(nx, ny));
                }
            }
        }
    }

    static void printAnswer() {
        if (canEnd) {
            System.out.println(visited[endX][endY]);
            return;
        }
        System.out.println("KAKTUS");
    }

    public static void main(String[] args) throws Exception {
        setInputs();
        bfs();
        printAnswer();
    }
}

File: 백준/Gold/3055. 탈출/탈출.py
from collections import deque
import sys
input = sys.stdin.readline


m, n = map(int, input().split())
graph = [list(input().strip()) for _ in range(m)]
visited = [[0] * n for _ in range(m)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

q = deque()

for i in range(m):
    for j in range(n):
        if graph[i][j] == '*':
            q.appendleft((i, j))
        elif graph[i][j] == 'S':
            q.append((i, j))
        elif graph[i][j] == 'D':
            end_x, end_y = i, j


can_end = False
while q:
    x, y = q.popleft()
    if graph[end_x][end_y] == 'S':
        can_end = True
        break

    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]

        if nx < 0 or nx >= m or ny < 0 or ny >= n:
            continue

        # 고슴도치가 다음칸으로 이동할 때 . or D라면 / 이동이 가능하다.
        if graph[x][y] == 'S' and (graph[nx][ny] == '.' or graph[nx][ny] == 'D'):
            graph[nx][ny] = 'S'
            visited[nx][ny] = visited[x][y] + 1
            q.append((nx, ny))

        # 물이 다음칸으로 이동할 때 . or S라면 / 이동이 가능하다.
        # 물이 먼저 움직인다고 가정했을 때 물이 이동하는 방향이 S일 때는 고려안해도된다.
        if graph[x][y] == '*' and (graph[nx][ny] == '.'):
            graph[nx][ny] = '*'
            q.append((nx, ny))

if can_end:
    print(visited[end_x][end_y])
else:
    print("KAKTUS")
